### ISSUE TITLE
`SubFigure` legend not included in figure bounding box

### ISSUE THREAD

**Issue by tomneep** [matplotlib/matplotlib#20736]
### Bug report

**Bug summary**

This relies on #20723 being addressed. 

As reported in [a comment on that issue](https://github.com/matplotlib/matplotlib/issues/20723#issuecomment-886066207), a legend added to a `SubFigure` ~~is not included when calculating the bbox of the parent figure and so the subfigure legend~~ is cut off the plot when saving a figure with `bbox_inches="tight"`.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
fig = plt.figure()
subfig = fig.subfigures()
ax = subfig.subplots()
ax.plot([0, 1, 2], [0, 1, 2], label="test")
subfig.legend()
print(fig.get_default_bbox_extra_artists())
plt.savefig("test.png", bbox_inches="tight")
```

**Actual outcome**

![test](https://user-images.githubusercontent.com/17545360/126876433-d8ee4d98-501e-4800-8b9a-1390c4faf0c1.png)

```python
[<matplotlib.figure.SubFigure at 0x7fc17864e700>,
 <matplotlib.lines.Line2D at 0x7fc17874aa90>,
 <matplotlib.spines.Spine at 0x7fc178612700>,
 <matplotlib.spines.Spine at 0x7fc178612910>,
 <matplotlib.spines.Spine at 0x7fc178612c70>,
 <matplotlib.spines.Spine at 0x7fc17864e5e0>,
 <matplotlib.axis.XAxis at 0x7fc178612c10>,
 <matplotlib.axis.YAxis at 0x7fc17864ed60>,
 <matplotlib.patches.Rectangle at 0x7fc178738730>]
```

**Expected outcome**

The subfigure legend should be included in the bbox. 

![test](https://user-images.githubusercontent.com/17545360/126876497-1e4c1a0e-f6ab-4608-b0aa-5f35f32e00a3.png)

**Matplotlib version**

  * Operating system: Mac OS 11.4
  * Matplotlib version (`import matplotlib; print(matplotlib.__version__)`): 3.4.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg
  * Python version: 3.8.3

**Comment by jklymak** [comment]
I think the Figure's extra artists should only be the SubFigures, and anything added at the Figure level.  I don't think things at the SubFigure level should bubble up to the Figure level.  

OTOH there are issues here that will require some thought.  Folks enjoy adding legends _outside_ the figure, and then letting bbox_inches='tight' expand the figure to match.  Its hard to know how that will be compatible with SubFigure.

**Comment by tomneep** [comment]
> Folks enjoy adding legends _outside_ the figure, and then letting bbox_inches='tight' expand the figure to match. Its hard to know how that will be compatible with SubFigure.

Yes in fact this is exactly how I ran into this problem. I wanted to add a legend outside the axes of the subfigure.

**Comment by jklymak** [comment]
Yeah, that is my main point - Figure and SubFigure should be different from one another (probably).

**Comment by tomneep** [comment]
I had a quick investigation and ignoring the fact that `fig.get_default_bbox_extra_artists()` includes stuff it perhaps shouldn't, I think the problem seems to be that `FigureBase.get_tightbbox()` returns the bounding box in figure inches. I think this is probably wrong for a `SubFigure.` As a quick test, I just added a `return` in `FigureBase.get_tightbbox()` here

https://github.com/matplotlib/matplotlib/blob/d2a97105aee4d892f7ba854f0531069157ab9b1a/lib/matplotlib/figure.py#L1642-L1646

such that `_bbox` is returned (instead of `bbox_inches`) if `self` is a `SubFigure`. This at least keeps the legend in the figure but breaks `test_subfigure_tightbbox` so some more investigation is needed. I'm not proposing this as the solution but hopefully this give some hint as to why even though the `SubFigure`  is in the list from `fig.get_default_bbox_extra_artists()` we don't seem to get the right bbox.

**Comment by oscargus** [comment]
Executing the code on current main branch leads to:
```
[<matplotlib.figure.SubFigure object at 0x00000185331DE340>,
<matplotlib.spines.Spine object at 0x00000185331DEA60>,
<matplotlib.spines.Spine object at 0x00000185331DEB50>,
<matplotlib.spines.Spine object at 0x00000185331DEC40>,
<matplotlib.spines.Spine object at 0x00000185331DED30>,
<matplotlib.patches.Rectangle object at 0x000001853321F5E0>]
```
so the line is no longer there. On the other hand, the resulting image looks like:
![test](https://user-images.githubusercontent.com/8114497/154823315-5d9a99f3-075a-4020-9eec-cb270faf3ae2.png)

**Comment by tomneep** [comment]
Thanks @oscargus. It looks to me like this was fixed by #21845 and can be closed.
